File: atcoder/abc262/d.cpp
#include <stdio.h>
const int MOD = 998244353;
const int MAX = 101;

int a[MAX];

int main() {
    int n;
    scanf("%d", &n);
    for(int i = 0; i < n; ++ i) {
        scanf("%d", &a[i]);
    }
    int tot = 0;

    for (int m = 1; m <= n; ++ m) {
        int dp[n+1][n+1][n];
        for(int i = 0; i < n+1; ++ i) {
            for(int j = 0; j < n+1; ++ j) {
                for(int k = 0; k < n; ++ k) {
                    dp[i][j][k] = 0;
                }
            }
        }
        dp[0][0][0] = 1;

        for(int i = 0; i < n; ++ i) {
            for(int j = 0; j < n; ++ j) {
                for(int k = 0; k < n; ++ k){
                    dp[i+1][j+1][(k+a[i])%m] = (dp[i+1][j+1][(k+a[i])%m] + dp[i][j][k]) % MOD;
                    dp[i+1][j][k] = (dp[i+1][j][k] + dp[i][j][k]) % MOD;
                }
            }
        }

        tot = (tot + dp[n][m][0]) % MOD;
        
    }
    printf("%d\n", (tot%MOD+MOD)%MOD);
}

File: atcoder/abc262/d.py
import random

def crange(start, end, step=1):
    dir = 1 if start < end else -1
    if start > end and step > 0: step = -step
    return range(start, end + dir, step)


# binomial
MOD = 998244353
class Binomial:
    def __init__(self, n):
        n = min(n, MOD)
        self.fact = [1, 1]
        self.inv_fact = [1, 1]
        self.inv = [0, 1]
        
        for i in crange(2, n):
            self.fact.append(self.fact[-1] * i % MOD)
            self.inv.append((MOD - MOD // i) * self.inv[MOD % i] % MOD)
            self.inv_fact.append(self.inv_fact[-1] * self.inv[-1] % MOD)
    

    def comb(self, n, m):
        if m < 0 or m > n:
            return 0
        m = min(m, n-m)
        return self.fact[n] * self.inv_fact[m] * self.inv_fact[n-m] % MOD

    ''' Lucas theorem
    - mod should be less than 1e5
    '''
    def lucas(self, n, m): 
        if m == 0:
            return 1
        return self.comb(n % MOD, m % MOD) * \
            self.lucas(n // MOD, m // MOD) % MOD


class Encodict:
    def __init__(self, func=lambda : 0):
        self.RANDOM = random.randint(0, 1<<32)
        self.default = func
        self.dict = {}
    
    def __getitem__(self, key):
        k = self.RANDOM ^ key
        if k not in self.dict:
            self.dict[k] = self.default()
        return self.dict[k]
    
    def __setitem__(self, key, item):
        k = self.RANDOM ^ key
        self.dict[k] = item

    def keys(self):
        return [self.RANDOM ^ i for i in self.dict]
    
    def items(self):
        return [(self.RANDOM ^ i, self.dict[i]) for i in self.dict]
    
    def sorted(self, by_value=False, reverse=False):
        if by_value:
            self.dict = dict(sorted(self.dict.items(), \
                key=lambda x:x[1], reverse=reverse))
        else:
            self.dict = dict(sorted(self.dict.items(), \
                key=lambda x:self.RANDOM^x[0], reverse=reverse))

n = int(input())
a = list(map(int, input().split()))

d = Encodict(lambda : 0)

l = [[] for i in range(n+1)]

for x in a:
    d[x] += 1
    for i in range(1, n+1):
        if x % i == 0:
            l[i].append(x)


tot = 0
bi = Binomial(n+1)
for i in range(1, n+1):
    tot = (tot + bi.comb(len(l[i]), i)) % MOD
    for x in d.keys():
        if d[x] >= i and x % i != 0:
            tot = (tot + bi.comb(d[x], i)) % MOD
            

print(tot)



